core/templates/pages/exploration-player-page/services/stats-reporting.service.ts

```diff
@@ -16,344 +16,416 @@
 
-        postStatsToBackend();
-        PlaythroughService.recordExplorationQuitAction(
-          stateName, stateStopwatch.getTimeInSecs());
+      this.postStatsToBackend();
+      this.playthroughService.recordExplorationQuitAction(
+        stateName, this.stateStopwatch.getTimeInSecs());
 
-        PlaythroughService.recordPlaythrough(true);
-        explorationIsComplete = true;
-      },
-      recordAnswerSubmitted: function(
-          stateName, params, answer, answerGroupIndex, ruleIndex,
-          classificationCategorization, feedbackIsUseful) {
-        if (!aggregatedStats.state_stats_mapping.hasOwnProperty(stateName)) {
-          createDefaultStateStatsMapping(stateName);
-        }
-        aggregatedStats.state_stats_mapping[stateName].total_answers_count += 1;
-        if (feedbackIsUseful) {
-          aggregatedStats.state_stats_mapping[
-            stateName].useful_feedback_count += 1;
-        }
-        currentStateName = stateName;
-        $http.post(getFullStatsUrl('ANSWER_SUBMITTED'), {
-          answer: answer,
-          params: params,
-          version: explorationVersion,
-          session_id: sessionId,
-          client_time_spent_in_secs: stateStopwatch.getTimeInSecs(),
-          old_state_name: stateName,
-          answer_group_index: answerGroupIndex,
-          rule_spec_index: ruleIndex,
-          classification_categorization: classificationCategorization
-        });
-      },
-      recordMaybeLeaveEvent: function(stateName, params) {
-        currentStateName = stateName;
-        $http.post(getFullStatsUrl('EXPLORATION_MAYBE_LEFT'), {
-          client_time_spent_in_secs: stateStopwatch.getTimeInSecs(),
-          collection_id: optionalCollectionId,
-          params: params,
-          session_id: sessionId,
-          state_name: stateName,
-          version: explorationVersion
-        });
+      this.playthroughService.recordPlaythrough(true);
+      this.explorationIsComplete = true;
+    }
 
-        postStatsToBackend();
-
-        PlaythroughService.recordExplorationQuitAction(
-          stateName, stateStopwatch.getTimeInSecs());
-        PlaythroughService.recordPlaythrough();
-      },
-      recordAnswerSubmitAction: function(
-          stateName, destStateName, interactionId, answer, feedback) {
-        PlaythroughService.recordAnswerSubmitAction(
-          stateName, destStateName, interactionId, answer, feedback,
-          stateStopwatch.getTimeInSecs());
+    // The type of params is declared as Object since it can vary depending
+    // on the stateName.
+    recordAnswerSubmitted(
+        stateName: string, params: Object, answer: string,
+        answerGroupIndex: number, ruleIndex: number,
+        classificationCategorization: string, feedbackIsUseful: boolean): void {
+      if (!this.aggregatedStats.state_stats_mapping.hasOwnProperty(stateName)) {
+        this.createDefaultStateStatsMapping(stateName);
```

Comment:
Let's just move the check for pre-existing stats into `createDefaultStateStatsMapping` instead? Then this can simply become something like:

```typescript
this.createDefaultStateStatsMappingIfMissing(stateName);
````

Reply:
Done.